File: src/wrsn/Individual.java
package wrsn;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Random;


import utils.Factor;
import utils.Factory;


public class Individual {
    private static Random rd = new Random();
    private int N; // so luong gene
    private ArrayList<Integer> path; // duong di cua xe sac qua cac sensor

    private double totalDistance;   // tong chi phi duong di
    private ArrayList<Double> taus; // thoi gian sac cua cac sensor
    private double fitnessF;    // fitness f - danh gia duong di
    private int fitnessG;    // fitness g - danh gia thoi gian sac

    public Individual(Map map) {
        N = map.getN();
        path = new ArrayList<>();
        taus = new ArrayList<>();
        // Chi gan gia tri cho duong di truoc
        for (int i = 0; i < N; i++) {
            setNode(i, i);
        }
        // Sinh hoan vi
        Collections.shuffle(path);
    }

    public Individual(Individual individual) {
        N = individual.getN();
        path = new ArrayList<>(individual.getPath());
        taus = new ArrayList<>(individual.getTaus());
    }
    
    // for phase 2
    public Individual(Map map, Individual individual) {
        HashMap<Double, Boolean> isDuplicate  = new HashMap<>();
        taus = new ArrayList<>();
        N = individual.getN();
        path = new ArrayList<>(individual.getPath());

        
        this.calculateTotalDistance(map);
		double E_T = totalDistance * WCE.P_M / WCE.V;
        double t = (WCE.E_MC - E_T)/WCE.U;
        ArrayList<Double> max_t_list = new ArrayList<>();
		for(int i = 0; i < N; i++){
			double max_t = (Sensor.E_MAX - Sensor.E_MIN) / (WCE.U - map.getSensor(i).getP());
			max_t_list.add(max_t / t);
		}
        
        ArrayList<Double> t_sensor_charged = new ArrayList<>();


// Chia thời gian ra N khoảng 
        t_sensor_charged.add(0.0);
		for (int i = 0 ; i < N - 1  ; i++) {
            Double tmp = rd.nextDouble();
            while( (tmp==0 || isDuplicate.containsKey(tmp))) {
                tmp = rd.nextDouble();
            }
            t_sensor_charged.add(tmp);
            isDuplicate.put(tmp, true);
        }
        t_sensor_charged.add(1.0);
        Collections.sort(t_sensor_charged);
        
        // Trường hợp sensor cuối dư t_charged > t_max thì đem chia lại cho các sensor khác
        boolean flag=false;
        for (int i = 1 ; i<t_sensor_charged.size() ; i++) {
            Double t_sensor = t_sensor_charged.get(i)-t_sensor_charged.get(i-1);
            if (i==t_sensor_charged.size()-1 && t_sensor>max_t_list.get(i-1)) {
                flag=true;
            }

            if (t_sensor > max_t_list.get(i-1)) {
                double current = t_sensor_charged.get(i);
                t_sensor_charged.set(i-1,current - (t_sensor-max_t_list.get(i-1))); 
            }
        }

        if (flag) {
            for (int i = t_sensor_charged.size()-2 ; i>=0 ; i--) {
                Double t_sensor = t_sensor_charged.get(i+1)-t_sensor_charged.get(i);
                if (t_sensor > max_t_list.get(i)) {
                    Double current = t_sensor_charged.get(i);
                    t_sensor_charged.set(i,current + (t_sensor-max_t_list.get(i))); 
                    }
                else {
                    break;
                }
            }
        }

        for (int i = 1 ; i<t_sensor_charged.size() ; i++) {
            taus.add(t_sensor_charged.get(i)-t_sensor_charged.get(i-1));
        }
    }

    public void calculateTotalDistance(Map map) {
        totalDistance = 0;
        int nSensors = path.size();
        for (int i = 0; i < nSensors; i++) {
            int previous = i == 0 ? Factor.BS_INDEX : path.get(i - 1);
            int current = path.get(i);
            totalDistance += map.distanceCalculate(previous, current);
        }
        totalDistance += map.distanceCalculate(path.get(nSensors - 1), Factor.BS_INDEX);
    }
    public void calculateFitnessFGACS(Map map) {
        Factory factory = new Factory();
        fitnessF  = factory.fitnessFGACS(path, map);
    }
    public void calculateFitnessGGACS(Map map) {
        Factory factory = new Factory();
        fitnessG  = factory.fitnessGGACS(this, map);
    }
    private void calculateFitnessF(Map map) {
        Factory factory = new Factory();
        fitnessF  = factory.fitnessF(path, map);
    }

    public double getTotalDistance() {
        return totalDistance;
    }

    public int getN() {
        return N;
    }

    public double getFitnessF() {
        return fitnessF;
    }

    public int getFitnessG() {
        return fitnessG;
    }

    public ArrayList<Integer> getPath() {
        return path;
    }

    public void setPath(ArrayList<Integer> path) {
        this.path = path;
    }

    public int getNode(int index) {
        return path.get(index);
    }

    public void setNode(int index, int node) {
        if (index > path.size() - 1) path.add(node);
        else path.set(index, node);
    }

    public ArrayList<Double> getTaus() {
        return taus;
    }

    public void setTaus(ArrayList<Double> taus) {
        this.taus = taus;
    }

    public Double getTau(int index) {
        return taus.get(index);
    }

    public void setTau(int index, Double tau) {
        if (index > taus.size() - 1) taus.add(tau);
        else taus.set(index, tau);
    }
}

File: wrsn/individual.py
import os
import random
import numpy as np

from utils import factor
from utils import factory

from wrsn.map import Map
from wrsn.individual import Individual
from wrsn.wce import WCE
from wrsn.sensor import Sensor


class Individual:
    def __init__(self, map=None, individual: Individual=None):
        self.N = 0 # So luong gene
        self.path = [] # duong di cua xe sac qua cac sensor
        self.taus = [] # thoi gian sac cua cac sensor
        self.totalDistance = 0 # tong chi phi duong di
        self.fitnessF = 0 # fitness danh gia duong di
        self.fitnessG = 0 # fitness danh gia thoi gian sac

        self.WCE_ = WCE()
        self.Sensor_ = Sensor()

        if (map != None) & (individual == None): 
            self.N = map.getN()
            # Chi gan gia tri cho duong di truoc
            for i in range(self.N):
                self.setNode(i, i)
            # Sinh hoan vi
            random.shuffle(self.path)
        elif (map == None) & (individual != None):
            self.N = individual.getN()
            self.path = individual.getPath()
            self.taus = individual.getTaus()
        # Phase 2
        else:
            self.isDuplicate = {}
            self.N = individual.getN()
            self.path = individual.getPath()
            self.taus = []

            self.calculateTotalDistance(map)
            self.E_T = self.totalDistance * self.WCE_.P_M / self.WCE_.V
            t = (self.WCE_.E_MC - self.E_T) / self.WCE_.U

            self.max_t_list = []
            for i in range(self.N):
                max_t = (self.Sensor_.E_MAX - self.Sensor_.E_MIN) / (self.WCE_.U - map.getSensor(i).getP())
                self.max_t_list.append(max_t / t)

            self.t_sensor_charged = []
        
            # Chia thoi gian ra n khoang
            self.t_sensor_charged.append(0.0)
            for i in range(self.N - 1):
                tmp = np.random.randn()
                while (tmp == 0 or tmp in self.isDuplicate.keys()):
                    tmp = np.random.randn()
                self.t_sensor_charged.append(tmp)
                self.isDuplicate[tmp] = True
            
            self.t_sensor_charged.append(1.0)
            self.t_sensor_charged.sort()

            # Trường hợp sensor cuối dư t_charged > t_max 
            # thì đem chia lại cho các sensor khác
            flag = False
            for i in range(1, len(self.t_sensor_charged)):
                t_sensor = self.t_sensor_charged[i] - self.t_sensor_charged[i - 1]
                if (i == len(self.t_sensor_charged)) & (t_sensor > self.max_t_list[i - 1]):
                    flag = True
                
                if t_sensor > self.max_t_list[i - 1]:
                    current = self.t_sensor_charged[i]
                    self.t_sensor_charged[i-1] = current - (t_sensor - self.max_t_list[i - 1])

            if flag:
                for i in range(len(self.t_sensor_charged) - 2, -1, -1):
                    t_sensor = self.t_sensor_charged[i + 1] - self.t_sensor_charged[i]
                    if t_sensor > self.max_t_list[i]:
                        current = self.t_sensor_charged[i]
                        self.t_sensor_charged[i] = current + (t_sensor - self.max_t_list[i])
                    else:
                        break
            
            for i in range(1, len(self.t_sensor_charged)):
                self.taus.append(self.t_sensor_charged[i] - self.t_sensor_charged[i - 1])

    def calculateTotalDistance(self, map: Map):
        nSensors = len(self.path)
        for i in range(nSensors):
            if i == 0:
                previous = factor.BS_INDEX
            else:
                previous = self.path[i - 1]
            current = self.path[i]
            self.totalDistance += map.distanceCalculate(previous, current)
        
        self.totalDistance += map.distanceCalculate(self.path[nSensors - 1], factor.BS_INDEX)

    def calculateFitnessFGACS(self, map):
        pass

    def calculateFitnessF(self, map):
        pass

    def getTotalDistance(self):
        return self.totalDistance

    def getN(self):
        return self.N

    def getFitnessF(self):
        return self.fitnessF

    def getFitnessG(self):
        return self.fitnessG

    def getPath(self):
        return self.path

    def setPath(self, path):
        self.path = path

    def getNode(self, index):
        return self.path[index]

    def setNode(self, index, node):
        if index > len(self.path) - 1:
            self.path.append(node)
        else:
            self.path[index] = node
    
    def getTaus(self):
        return self.taus

    def setTaus(self, taus):
        self.taus = taus
    
    def getTau(self, index):
        return self.taus[index]
    
    def setTau(self, index, tau):
        if index > len(self.taus - 1):
            self.taus.append(tau)
        else:
            self.taus[index] = tau